### PR TITLE
Add validation for the chronological order of conference dates

### LINKED ISSUE #503: Conferences - small improvements for adding new
- [x] 1/ When trying to add a conference without its name, conference is not being added, but the view is broken. I guess there should be for example some alert like 'fill in the name' or a view informing that a conference haven't been added with "Back to index" option.
  screenshot from 2016-07-15 16-00-16
- [ ] 2/ It's possible to add a conference with end date earlier than start date.
  screenshot from 2016-07-15 15-42-21
  screenshot from 2016-07-15 15-44-37
- [x] 3/ When adding a conference it's possible to choose not round 1 or 2, but 16, 21, -11 etc. You get a view like conference has been created /but it's not on the list/. There's no alert about incorrect round chosen. I think there should be only two values to chose from /1 or 2/.
- [x] 4/ Similar problem in tickets and flights. You can create a conf with -50 tickets and -40 flight and it gets on the list.
  screenshot from 2016-07-15 16-55-35
  screenshot from 2016-07-15 17-00-40

These are not crucial improvements, as conferences are being added by Orga Team and so risk of mistake is close to zero. Only suggesting.

### PR DESCRIPTION
**Pull request by F3PiX**
Closes #503 (bullet point 2)
Replacement for #552 

- Add custom validator method + specs + factory
- Minor cosmetics

### CODE REVIEW

File: spec/controllers/orga/conferences_controller_spec.rb [hunk 1]
```diff
@@ -3,6 +3,8 @@
 RSpec.describe Orga::ConferencesController do
   render_views
 
+  it_behaves_like 'redirects for non-admins'
+
```
Comment (F3PiX) [inline spec/controllers/orga/conferences_controller_spec.rb]: Okay to add this?

File: spec/helpers/application_helper_spec.rb [hunk 2]
```diff
@@ -115,14 +115,14 @@
...
       end
 
       context 'with location' do
         let(:conference) { build_stubbed :conference, location: 'Melée Island, Carribean' }
 
-        it 'will add the location in brackets' do
+        xit 'will add the location in brackets' do
           link_text = "#{conference.name} (Melée Island, Carribean)"
           expect(subject).to eql [link_to(link_text, conference)]
         end
```
Comment (F3PiX) [inline spec/helpers/application_helper_spec.rb]: These tests are related to formatting the location and the dates in several views, handling the layout of conferences with and without info with dates.  
Now with the `presence of dates validations`, is that code still relevant? (For example, existing records.) Or should it be removed.

File: spec/factories/conferences.rb [hunk 3]
```diff
@@ -1,7 +1,9 @@
 FactoryGirl.define do
   factory :conference do
     name { [FFaker::CheesyLingo.title, 'Conf'].join ' ' }
-    tickets 2
```
Comment (F3PiX) [inline spec/factories/conferences.rb]: I removed the tickets, because they are not a requirement for validation. Plus, the only test that refers to the ticket attributes, overrides it.
(Copied from old PR)

File: app/controllers/orga/conferences_controller.rb [hunk 4]
```diff
@@ -35,9 +38,9 @@ def conference
...
     params[:conference] ? params.require(:conference).permit(
-      :name, :url, :location, :twitter, :tickets, :flights, :accomodation,
-      :'starts_on(1i)', :'starts_on(2i)', :'starts_on(3i)', :round,
-      :'ends_on(1i)', :'ends_on(2i)', :'ends_on(3i)', :lightningtalkslots,
+      :name, :url, :location, :twitter,
+      :tickets, :flights, :accomodation,
+      :starts_on, :ends_on, :round, :lightningtalkslots,
       attendances_attributes: [:id, :github_handle, :_destroy]
     ) : {}
   end
```
Comment (F3PiX) [inline app/controllers/orga/conferences_controller.rb]: (Regrouped and reformatted for better readibility.)

**Comment by F3PiX** [comment]
Re: my refactor fest. I'd really like to refactor the Conference's `(:starts_on, :ends_on)` to a DateRange object. I think the correct order is do the refactoring first, and the new validation afterwards. 
So, two questions: 1) Do you agree with the refactoring itself?
2) If so, should I do that first?

**Comment by klappradla** [comment]
On refactoring vs. feature first: do as you wish ✌️ 
It totally depends on the state of the code base and how easy it is for you to work with its current state. The "classic" approach is feature first, then refactoring. But of course, if that part of the codebase has not been touched in a while or whatever other reason one could find, it may of course also be way more efficient to do the refactoring first.

Sometimes you even know upfront, that you cannot use any of the existing code - it makes of course sense to just throw it all away and implement your new way of doing things from scratch then.

**TL;DR** do however it suits your workflow to best archive what you want to.

File: spec/helpers/application_helper_spec.rb (hunk 2 above)
Comment (klappradla) [inline spec/helpers/application_helper_spec.rb]: If the view helper is no longer used, yes, remove both, the helper and the tests. If it is still used, I don't see a point in removing the tests for it.

File: spec/models/conference_spec.rb [hunk 5]
```diff
@@ -7,6 +7,22 @@
   it { is_expected.to have_many(:attendees) }
   it { is_expected.to validate_presence_of(:name) }
   it { is_expected.to validate_presence_of(:round) }
+  it { is_expected.to validate_presence_of(:starts_on) }
+  it { is_expected.to validate_presence_of(:ends_on) }
+
+  describe 'it checks the chronological order of dates' do
+    subject { FactoryGirl.build(:conference) }
```
Comment (klappradla) [inline spec/models/conference_spec.rb]: It's a common best practise and in my opinion also increases readability to leave a blank line between `subject` and the first example.

I'm also in general not sure about the use of `subject` here? I'm not working with Rails on a day to day basis, but from what I remember, the `errors` field is only appended to a model and updated accordingly if `model.valid?` is called... So I'm not sure if your expectation here actually does what it should.

File: spec/models/conference_spec.rb [hunk 6]
```diff
@@ -7,6 +7,22 @@
...
+    it 'raises an error with an incorrect order of dates' do
+      subject.starts_on = '2017-07-15'
+      subject.ends_on = '2017-07-07'
+      expect(subject).to have(1).error_on(:ends_on)
```
Comment (klappradla) [inline spec/models/conference_spec.rb]: Personal preference, but I'm not a huge fan of the `rspec-collection-matchers` gem (I would love to remove it from the Teams App 😉  ) since it also continues to ship matchers from RSpec < v3, such as `errors_on`... In my opinion there's probably a reason why this is no longer in RSpec.

How about just using plain RSpec here?
```ruby
subject.valid?
expect(subject.errors[:ends_on]).not_to be_empty
```
Shouldn't it do the same?

File: spec/models/conference_spec.rb (hunk 5 above)
Comment (carpodaster) [inline spec/models/conference_spec.rb]: IIRC,  the has errors matcher internally calls `#valid?`

Comment (klappradla) [inline spec/models/conference_spec.rb]: Ah, would make sense @carpodaster, thanx 🙏

File: spec/helpers/application_helper_spec.rb (hunk 2 above)
Comment (F3PiX) [inline spec/helpers/application_helper_spec.rb]: Which reminds me: are there currently records of conferences without the dates set?
I don't have access to the database. 

(As far as I can see, the input form that prepopulate the dates was present [from the beginning of the Conferences model  ](https://github.com/rails-girls-summer-of-code/rgsoc-teams/commit/9bf58943bb7bc49c52ea2569e25962b5bf81c974#diff-da8e2af5531074e5eb6389d13add128a) )

File: app/controllers/orga/conferences_controller.rb [hunk 7]
```diff
@@ -35,9 +38,9 @@ def conference
 
   def conference_params
     params[:conference] ? params.require(:conference).permit(
```
Comment (F3PiX) [inline app/controllers/orga/conferences_controller.rb]: Also ancient code. I am not sure when I should and shouldn't touch existing code that is not explicitly part of the issue I am working on. 
I really _like_ the concept of leaving code cleaner than you found it, but isn't that the opposite of doing one thing at the time? (Refactor hat or Feature hat, IIRC) . 
How do I know when to clean up, or when to leave it alone?

File: spec/helpers/application_helper_spec.rb (hunk 2 above)
Comment (klappradla) [inline spec/helpers/application_helper_spec.rb]: I had a quick look into the current production data - could not find any conferences with `starts_on` or `ends_on` set to `nil`. So I guess that's not a problem...

File: app/controllers/orga/conferences_controller.rb (hunk 7 above)
Comment (klappradla) [inline app/controllers/orga/conferences_controller.rb]: You usually know it when you remove it, run the tests and everything still works 😉  But yeah, risky, let's just focus on what you actually want to change.

I also had a quick look at the whole controller code. There's this line:
```ruby
@conference ||= params[:id] ? Conference.find(params[:id]) : Conference.new(conference_params)
```
This for instance causes looking for a non-existing conference to just create a new one instead of raising an error (Record not Found) - so that's why the hack with the empty hash exists 💡

File: spec/models/conference_spec.rb (hunk 5 above)
Comment (F3PiX) [inline spec/models/conference_spec.rb]: :thinking: It looks is doesn't after all. 
Without `subject.valid?` `subject.errors[:ends_on]` is always empty. 
I can either add `subject.valid?`, or do 
`expect(subject).not_to be_valid`. Does it matter?

File: spec/helpers/application_helper_spec.rb (hunk 2 above)
Comment (F3PiX) [inline spec/helpers/application_helper_spec.rb]: Thanks for checking 👍

File: spec/factories/conferences.rb [hunk 8]
```diff
@@ -1,7 +1,9 @@
...
+    starts_on { Time.utc(Date.today.year, 7, 7) }
```
Comment (F3PiX) [inline spec/factories/conferences.rb]: (Noticed the CheesyLingo lib in the FFaker gem ^ ? That was created by Team Cheesy during Soc15, and dedicated to SoC.) #thingsrailsgirlsdo

**Comment by F3PiX** [comment]
I leave the daterange refactor for after this one is done.
Re: Removing/simplifying the application helper method: see the [commit message](https://github.com/rails-girls-summer-of-code/rgsoc-teams/pull/747/commits/de7333c4e79d84f8cc1e7191c37f34bccbb48e95)

**Comment by F3PiX** [comment]
new application helper

File: spec/models/conference_spec.rb (hunk 5 above)
Comment (klappradla) [inline spec/models/conference_spec.rb]:
```ruby
expect(subject).not_to be_valid
```
would generally be fine, if you can make sure that the error is not triggered from a completely different validation. So if doing it like this maybe also quickly check if the subject otherwise is valid

```ruby
subject { thing } 

it { is_expected.to be_valid }

# more specs
```

File: db/seeds.rb [hunk 9]
```diff
@@ -6,7 +6,7 @@
 FactoryGirl.create(:team, :last_season, kind: "sponsored")
 FactoryGirl.create(:team, :last_season, kind: "voluntary")
 
-# Users with different roles
+# # Users with different roles
```
Comment (klappradla) [inline db/seeds.rb]: Please remove the extra '#' *(or tell me what's the purpose of this 😉 )*

File: db/seeds.rb [hunk 10]
```diff
@@ -16,7 +16,7 @@
 FactoryGirl.create(:mentor)
 FactoryGirl.create(:supervisor)
 
-# To explore use cases where user has no role yet
+# # To explore use cases where user has no role yet
```
Comment (klappradla) [inline db/seeds.rb]: Again: Please remove the extra '#' *(or tell me what's the purpose of this 😉 )*

File: app/helpers/application_helper.rb [hunk 11]
```diff
@@ -67,10 +62,11 @@ def format_date(date, format = :long)
     date && l(date.to_date, format: format) || '-'
   end
 
+  ## Conferences
   def format_conference_date(starts_on, ends_on)
-    return unless [starts_on, ends_on].all?(&:present?)
-    starts_on = starts_on.strftime('%d %b %y')
-    ends_on = ends_on.strftime('%d %b %y')
+    return unless starts_on && ends_on
```
Comment (klappradla) [inline app/helpers/application_helper.rb]: Is this a scenario that could ever happen? As far as I understand it, there are presence validations on these fields already...?

File: app/views/orga/teams/show.html.slim [hunk 12]
```diff
@@ -79,7 +79,7 @@ nav.btn-group
   ul.conferences
     - @team.members.each do |user|
       - if user.conferences.in_current_season.any?
-        li = "#{link_to user.name, user}: #{links_to_conferences(user.conferences.in_current_season, true).join(',')}".html_safe
+        li = "#{link_to user.name, user}: #{links_to_conferences(user.conferences.in_current_season).join(',')}".html_safe
```
Comment (klappradla) [inline app/views/orga/teams/show.html.slim]: Since the helper method is already doing quite a lot, I would suggest to also move `join` and `html_safe` there. It looks kind of weird to me, having a rather long method called `links_to_conferences` not returning the actual links in a *"view-ready*" format.

What do you think?

It would also make this list item a bit cleaner:
```ruby
li = "#{link_to user.name, user}: #{link_to_conferences(user.conferencs.in_current_season)}
```

File: app/views/teams/show.html.slim [hunk 13]
```diff
@@ -91,7 +91,7 @@ nav.page
   ul.conferences
     - @team.members.each do |user|
       - if user.conferences.in_current_season.any?
-        li = "#{link_to user.name, user}: #{links_to_conferences(user.conferences.in_current_season, verbose: true).join(', ')}".html_safe
+        li = "#{link_to user.name, user}: #{links_to_conferences(user.conferences.in_current_season).join(', ')}".html_safe
```
Comment (klappradla) [inline app/views/teams/show.html.slim]: Same as for the other conferences link: I would suggest to move things into the helper if it's already there.

File: db/seeds.rb (hunk 9 above)
Comment (F3PiX) [inline db/seeds.rb]: It's a feature ;-) Like an H2, under H1 Users. But maybe MaudFlavoredMarkdown doesn't work that good if I have to explain it. Will delete it.

File: app/views/orga/teams/show.html.slim (hunk 12 above)
Comment (F3PiX) [inline app/views/orga/teams/show.html.slim]: I totally agreed, but I didn't find a way to solve it. Spent too much time on it, I'll make a note and will try it another time. Okay?

Comment (klappradla) [inline app/views/orga/teams/show.html.slim]: Sure ✌️ 

The easiest way would have been to just append the `join` etc. to the already existing method:
```ruby
def links_to_conferences(conferences)
  conferences.map do |conference|
    text = conference.name
    extra = [conference.location, format_conference_date(conference.starts_on, conference.ends_on)].reject(&:blank?).join(' – ')
    text += " (#{extra})" unless extra.blank?
    link_to(text, conference)
  end.join(',').html_safe
end
```
But yeah, there's probably a prettier way.

File: app/helpers/application_helper.rb [hunk 14]
```diff
@@ -96,13 +95,13 @@ def format_conference_twitter(twitter)
   def links_to_conferences(conferences)
     conferences.map do |conference|
       text = conference.name
-      extra = [conference.location, format_conference_date(conference.starts_on, conference.ends_on)].reject(&:blank?).join(' – ')
-      text += " (#{extra})" unless extra.blank?
+      details = [conference.location, format_conference_date(conference.starts_on, conference.ends_on)].reject(&:blank?).join(' – ')
+      text += " (#{details})"
       link_to(text, conference)
```
Comment (klappradla) [inline app/helpers/application_helper.rb]: Well, if the condition is gone now, I'd suggest to make this all a bit smaller:
```ruby
def links_to_conferences(conferences)
  details = [conference.location, format_conference_date(conference.starts_on, conference.ends_on)].reject(&:blank?).join(' – ')
  link_to "#{conference.name} (#{details})", conference
end
```

But well, optional 😉

File: app/views/orga/teams/show.html.slim (hunk 12 above)
Comment (F3PiX) [inline app/views/orga/teams/show.html.slim]: Yeah, I expected that to work too. And it didn't. That's how my quest started ;-)

File: app/helpers/application_helper.rb (hunk 14 above)
Comment (F3PiX) [inline app/helpers/application_helper.rb]: I hate it that I forgot that..... :-)
I'll merge on green. Thanks again a lot!!!
